File: src/main/kotlin/specscript/commands/datamanipulation/JsonPatch.kt
package specscript.commands.datamanipulation

import specscript.language.*
import specscript.util.applyPatch
import tools.jackson.databind.JsonNode
import tools.jackson.databind.node.ArrayNode
import tools.jackson.databind.node.ObjectNode

object JsonPatch : CommandHandler("Json patch", "core/data-manipulation"), ObjectHandler {

    override fun execute(data: ObjectNode, context: ScriptContext): JsonNode {

        val doc = data["doc"]
            ?: context.output
            ?: throw SpecScriptCommandError("Json patch needs 'doc' parameter or non-null output variable.")
        val patch = data.getParameter("patch") as ArrayNode

        return doc.applyPatch(patch)
    }
}

File: typescript/src/commands/json-patch.ts
import type { CommandHandler } from '../language/command-handler.js'
import type { JsonValue } from '../language/types.js'
import { isObject, isArray, SpecScriptCommandError, CommandFormatError } from '../language/types.js'
import type { ScriptContext } from '../language/context.js'
import { applyJsonPatch } from '../util/json-patch.js'

/**
 * Json patch: applies RFC 6902 JSON Patch operations to a document.
 */
export const JsonPatchCommand: CommandHandler = {
  name: 'Json patch',

  execute(data: JsonValue, context: ScriptContext): JsonValue {
    if (!isObject(data)) {
      throw new CommandFormatError('Json patch expects an object')
    }
    const doc = data['doc'] ?? context.output
    if (doc === undefined) {
      throw new SpecScriptCommandError(
        "Json patch needs 'doc' parameter or non-null output variable."
      )
    }
    const patch = data['patch']
    if (!isArray(patch)) {
      throw new CommandFormatError("Json patch requires a 'patch' array")
    }
    return applyJsonPatch(doc, patch)
  },
}
